include/modem/location.h

```diff
@@ -0,0 +1,326 @@
+		/**
+		 * A-GPS notification data frame used by GNSS to let the application know it
+		 * needs new assistance data, used with event LOCATION_EVT_GNSS_ASSISTANCE_REQUEST.
+		 */
+		struct nrf_modem_gnss_agps_data_frame request;
+#endif
+	};
+};
+
+/** GNSS configuration. */
+struct location_gnss_config {
+	/**
+	 * @brief Timeout (in seconds), meaning how long GNSS is allowed to run when trying to
+	 * acquire a fix. Zero means that the timer is disabled.
+	 *
+	 * @details Note that this is not real time as experienced by the user.
+	 * Since GNSS cannot run while LTE is operating, the running time is counted from
+	 * the instant when GNSS is allowed to start. Library waits until RRC connection
+	 * is inactive before starting GNSS. If LTE power saving mode (PSM) is enabled,
+	 * library waits until modem enters PSM before starting GNSS, thus maximizing
+	 * uninterrupted operating window and minimizing power consumption.
+	 */
+	uint16_t timeout;
+
+	/** Desired accuracy level. */
+	enum location_accuracy accuracy;
+
+	/**
+	 * @brief If accuracy is set to LOCATION_ACCURACY_HIGH, instead of using the first fix, GNSS
+	 * is allowed to run for a longer time and produce the configured number of fixes before the
+	 * library outputs the current location.
+	 *
+	 * @details This typically improves the location accuracy. If accuracy is set to
+	 * LOCATION_ACCURACY_NORMAL or LOCATION_ACCURACY_LOW this parameter has no effect.
+	 */
+	uint8_t num_consecutive_fixes;
+};
+
+/** LTE cellular positioning configuration. */
+struct location_cellular_config {
+	/**
+	 * @brief Timeout (in seconds) of how long the cellular positioning procedure can take.
+	 * Zero means that the timer is disabled.
+	 */
+	uint16_t timeout;
+
+	/** Used cellular positioning service. */
+	enum location_service service;
+};
+
+/** Wi-Fi positioning configuration. */
+struct location_wifi_config {
+	/**
+	 * @brief Timeout (in seconds) of how long the Wi-Fi positioning procedure can take.
+	 * Zero means that the timer is disabled.
+	 */
+	uint16_t timeout;
+
+	/** Used Wi-Fi positioning service. */
+	enum location_service service;
+};
+
+/** Positioning method configuration. */
+struct location_method_config {
+	/** Positioning method. */
+	enum location_method method;
+	union {
+		/** Configuration for LOCATION_METHOD_CELLULAR. */
+		struct location_cellular_config cellular;
+		/** Configuration for LOCATION_METHOD_GNSS. */
+		struct location_gnss_config gnss;
+		/** Configuration for LOCATION_METHOD_WIFI. */
+		struct location_wifi_config wifi;
+	};
+};
+
+/** Location request configuration. */
+struct location_config {
+	/** Number of location methods in 'methods'. */
+	uint8_t methods_count;
+	/**
+	 * @brief Selected positioning methods and associated configurations in priority order.
+	 *
+	 * @details Index 0 has the highest priority. Number of methods is indicated in
+	 * 'methods_count'.
+	 */
+	struct location_method_config *methods;
```

Comment:
Wouldn't it be easier to just have this as an array with a configurable amount of entries that have a range up to the maximum supported amount of services. I think we can spare the extra stack bytes.

Reply:
For what its worth. I think the user-experience should be taken into account here as well. Keeping it as an array makes the implementation a little bit easier (especially for super new users), and if the tradeoff is a trivial amount of bytes, I think its worth it.

Reply:
Option 3 is fine by me.

Reply:
Completely agree, simplicity of the API for users is very important. I don't fully know why we didn't do #3 before. We probably thought the config structures would grow much more than they seem to.